satellite/repair/checker/checker.go

```diff
@@ -107,53 +105,24 @@ func (checker *Checker) Close() error {
 func (checker *Checker) IdentifyInjuredSegments(ctx context.Context) (err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	err = checker.metainfo.Iterate(ctx, "", checker.lastChecked, true, false,
-		func(ctx context.Context, it storage.Iterator) error {
-			var item storage.ListItem
-
-			defer func() {
-				var nextItem storage.ListItem
-				it.Next(ctx, &nextItem)
-				// start at the next item in the next call
-				checker.lastChecked = nextItem.Key.String()
-				// if we have finished iterating, send and reset durability stats
-				if checker.lastChecked == "" {
-					// send durability stats
-					mon.IntVal("remote_files_checked").Observe(checker.monStats.remoteFilesChecked)
-					mon.IntVal("remote_segments_checked").Observe(checker.monStats.remoteSegmentsChecked)
-					mon.IntVal("remote_segments_needing_repair").Observe(checker.monStats.remoteSegmentsNeedingRepair)
-					mon.IntVal("remote_segments_lost").Observe(checker.monStats.remoteSegmentsLost)
-					mon.IntVal("remote_files_lost").Observe(int64(len(checker.monStats.remoteSegmentInfo)))
-
-					// reset durability stats for next iteration
-					checker.monStats = durabilityStats{}
-				}
-			}()
-
-			for it.Next(ctx, &item) {
-				pointer := &pb.Pointer{}
-
-				err = proto.Unmarshal(item.Value, pointer)
-				if err != nil {
-					return Error.New("error unmarshalling pointer %s", err)
-				}
-				remote := pointer.GetRemote()
-				if remote == nil {
-					continue
-				}
-
-				err = checker.updateSegmentStatus(ctx, pointer, item.Key.String(), &checker.monStats)
-				if err != nil {
-					return err
-				}
-			}
-			return nil
-		},
-	)
+	observer := &checkerObserver{
+		repairQueue: checker.repairQueue,
+		irrdb:       checker.irrdb,
+		nodestate:   checker.nodestate,
+		monStats:    durabilityStats{},
+		log:         checker.logger,
+	}
+	err = checker.metaLoop.Join(ctx, observer)
 	if err != nil {
-		return err
+		return Error.New("IdentifyInjuredSegments err %v", err)
```

Comment:
For now, there is no way to conserve the `err` stack adding some additional message unless we use `Combine`, however, I have the feeling that `Combine` is not for this kind of things, so we should log `err` before returning.

Assess if `Combine` is fine, if you think that it isn't appropriated then log `err`

Reply:
On the other hand, I'm not entirely sure that if there is an error we want to terminate these process because of the loop in https://github.com/storj/storj/pull/2620/files#diff-dc8ddcff2c5de4611c4f5d3eda6fd766L85 will terminate and the process will only run once the satellite is tear down an started again.

I'm not saying that this is a problem just here, it's also happening in `checker.IrreparableProcess` and likely in other processes that the satellite run in `satellite/peer.Run`.

We could open a ticket to find out where this is happening and adopt a strategy, for example:

1. The person on call get paged and it has to restart the satellite as soon as possible in order of having again these process running.
2. We don't stop the service but we report, the person on call is paged, then she/he can check the errors and decide if to restart or keep it in observation to see if the same error is happening again in the short time and make the decision of restarting.

Reply:
I don't think the additional messaging is important enough that we need to wrap it. Also, this error can't be wrapped anymore because it was causing a context canceled error to fail a lot of tests, since the satellite would not end nicely.

Regarding your other point, it seems to me like most services on the satellite will return an error that will crash the satellite, and it will need to be restarted by whoever is on call. I like the idea of reporting the issue when it occurs, but not stopping the service by returning an error, I just don't know how we would do that right now. And if we just log the error out, we might not notice it, especially if it does not happen often.

Reply:
> I don't think the additional messaging is important enough that we need to wrap it. Also, this error can't be wrapped anymore because it was causing a context canceled error to fail a lot of tests, since the satellite would not end nicely.

I don't get how that can fails, the original error is lost so if it's `context canceled` there is no possibility to check it.
No saying that it has to be wrapped, mostly I'm trying to understand the problem.

By the way, before I missed, that it could be `Error.Wrap(err)`, is that also failing?

> Regarding your other point, it seems to me like most services on the satellite will return an error that will crash the satellite, and it will need to be restarted by whoever is on call.

Yes, whatever option I've described above imply so.

> I like the idea of reporting the issue when it occurs,

That's the thing, if I'm not mistaken, right now it won't report it until the satellite finish due to the `errgroup.Wait()`, so we won't notice it.

> but not stopping the service by returning an error,

I haven't suggested that; it could be in that way if we would run more than one instance of the same satellite, but that isn't the case right now.

> I just don't know how we would do that right now. And if we just log the error out, we might not notice it, especially if it does not happen often.

We will notice less if there is nothing in the log reporting it (per my above comment).

We can return the error and stop the process as it's happening now but before returning it log the error.

Reply:
What I meant is that when we do `Error.New....` instead of just returning `err`, the satellite would not close nicely because of this in `satellite/peer.go`:
```
group.Go(func() error {
	return errs2.IgnoreCanceled(peer.Repair.Checker.Run(ctx))
})
```
The context canceled error is not properly ignored when we return `Error.New(...` which is why I just returned `err` directly.

Reply:
I can log out the errors before we return them, but I do not know when we should return, or log and then return.

Reply:
> The context canceled error is not properly ignored when we return Error.New(... which is why I just returned err directly.

Oh man! I was checking the changes where I found this conversation in "files changed" but it turns that they weren't the last changes so I was reading the same than before and I thought that you didn't do any change. I had to manually go to the line to find the new change.

It makes sense now what you were saying.

>  can log out the errors before we return them, but I do not know when we should return, or log and then return.

Currently, the easiest is logging it before returning it, but we have not to log it if it's context canceled using [`errs2.IsCanceled`](https://godoc.org/storj.io/storj/internal/errs2#IsCanceled)

Reply:
I checked if `!errs2.IsCanceled` and logged if it is not canceled. I plan to merge shortly unless anyone brings up other concerns